core/src/main/java/io/questdb/griffin/engine/groupby/SampleByFirstLastRecordCursorFactory.java

```diff
@@ -0,0 +1,879 @@
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.groupby;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.sql.*;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.SqlKeywords;
+import io.questdb.griffin.engine.EmptyTableRecordCursor;
+import io.questdb.griffin.model.ExpressionNode;
+import io.questdb.griffin.model.QueryColumn;
+import io.questdb.std.*;
+
+public class SampleByFirstLastRecordCursorFactory implements RecordCursorFactory {
+    private static final int FILTER_KEY_IS_NULL = 0;
+    private static final int ITEMS_PER_OUT_ARRAY_SHIFT = 2;
+    private static final int FIRST_OUT_INDEX = 0;
+    private static final int LAST_OUT_INDEX = 1;
+    private static final int TIMESTAMP_OUT_INDEX = 2;
+    private final RecordCursorFactory base;
+    private final GenericRecordMetadata groupByMetadata;
+    private final int[] firstLastIndexByCol;
+    private final int[] queryToFrameColumnMapping;
+    private final int pageSize;
+    private final int maxSamplePeriodSize;
+    private final SingleSymbolFilter symbolFilter;
+    private final int groupBySymbolColIndex;
+    private final int timestampIndex;
+    private final SampleByFirstLastRecordCursor sampleByFirstLastRecordCursor;
+    private DirectLongList rowIdOutAddress;
+    private DirectLongList samplePeriodAddress;
+    private DirectLongList crossFrameRow;
+    private int groupByTimestampIndex = -1;
+
+    public SampleByFirstLastRecordCursorFactory(
+            RecordCursorFactory base,
+            TimestampSampler timestampSampler,
+            GenericRecordMetadata groupByMetadata,
+            ObjList<QueryColumn> columns,
+            RecordMetadata metadata,
+            Function timezoneNameFunc,
+            int timezoneNameFuncPos,
+            Function offsetFunc,
+            int offsetFuncPos,
+            int timestampIndex,
+            SingleSymbolFilter symbolFilter,
+            int configPageSize
+    ) throws SqlException {
+        this.base = base;
+        this.groupBySymbolColIndex = symbolFilter.getColumnIndex();
+        this.queryToFrameColumnMapping = new int[columns.size()];
+        this.firstLastIndexByCol = new int[columns.size()];
+        this.crossFrameRow = new DirectLongList(columns.size());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *crossFrameRow*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).